### PR TITLE
Handle `match` with `pytest.raises()`

### PR DESCRIPTION
**Pull request by blueyed**
TODO:

- [x] changelog rephrasing

Fixes https://github.com/pytest-dev/pytest/issues/6752.

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus  #5529 as far as i can tell introduced the bug and i missed it back then

File: src/_pytest/python_api.py [hunk 1]
```diff
@@ -566,7 +565,6 @@ def raises(  # noqa: F811
 def raises(  # noqa: F811
     expected_exception: Union["Type[_E]", Tuple["Type[_E]", ...]],
     *args: Any,
-    match: Optional[Union[str, "Pattern"]] = None,
     **kwargs: Any
 ) -> Union["RaisesContext[_E]", Optional[_pytest._code.ExceptionInfo[_E]]]:
```
Comment (blueyed) [inline src/_pytest/python_api.py]: @bluetech do the typing changes make sense?

File: changelog/6752.bugfix.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Pass `match` keyword argument to the tested function with
```
Comment (bluetech) [inline changelog/6752.bugfix.rst]: Suggested rephrasing:

```
When :py:func`pytest.raises` is used as a function (as opposed to a context manager),
a `match` keyword argument is now passed through to the tested function. Previously
it was swallowed and ignored (regression in pytest 5.1.0).
```

File: src/_pytest/python_api.py (hunk 1 above)
Comment (bluetech) [reply src/_pytest/python_api.py]: Yes, the types look good to me.

Comment (RonnyPfannschmidt) [reply src/_pytest/python_api.py]: is there a way to type it so, context manager usage and function usage get the different parameter sets correct?

so no args, but matches typed for contextmanger usage, and args, kwargs when the first "arg" is a callable?

Comment (bluetech) [reply src/_pytest/python_api.py]: @RonnyPfannschmidt, I believe it already does that (unless I'm confused).

For callers, only the `@overloads` are used for type checking, and there are two overloads above, one for the context manager case and one for the function case.

The annotations on the function def without the `@overload` is just used for type checking the *implementation*. The implementation needs to handle both cases so the types use `Union` etc.

**Comment by blueyed** [comment]
Rebased and fixed the typing in 37a283b.

File: src/_pytest/python_api.py [hunk 3]
```diff
@@ -557,16 +557,14 @@ def raises(  # noqa: F811
...
+) -> _pytest._code.ExceptionInfo[_E]:
     ...  # pragma: no cover
 
 
 def raises(  # noqa: F811
     expected_exception: Union["Type[_E]", Tuple["Type[_E]", ...]],
     *args: Any,
-    match: Optional[Union[str, "Pattern"]] = None,
     **kwargs: Any
 ) -> Union["RaisesContext[_E]", Optional[_pytest._code.ExceptionInfo[_E]]]:
```
Comment (bluetech) [inline src/_pytest/python_api.py]: The `Optional` should be removed here too.

**Pull request merged**